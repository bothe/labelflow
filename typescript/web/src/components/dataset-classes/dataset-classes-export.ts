import { stringifyCsv } from "@labelflow/common-resolvers";
import {
  getDatasetClassesExportName,
  triggerClientDownload,
} from "../../utils";
import { LabelClassWithShortcut } from "./types";

export const generateCsv = (classes: LabelClassWithShortcut[]) => {
  const rows = classes.map((labelClass) => [labelClass.name]);
  return stringifyCsv(rows, { header: false });
};

export const exportDatasetClasses = async (
  datasetSlug: string,
  classes: LabelClassWithShortcut[]
) => {
  const csv = await generateCsv(classes);
  const csvFile = new Blob([csv], { type: "application/csv" });
  const exportName = getDatasetClassesExportName(datasetSlug);
  const fileName = `${exportName}.csv`;
  triggerClientDownload(csvFile, fileName);
};
