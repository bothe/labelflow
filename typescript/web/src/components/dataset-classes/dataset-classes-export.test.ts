import { waitFor } from "@testing-library/react";
import { exportDatasetClasses, generateCsv } from "./dataset-classes-export";
import {
  DATASET_SLUG,
  LABEL_CLASSES_DATA,
} from "./dataset-classes-export.fixtures";

describe("LabelClasses CSV export", () => {
  it("Generates proper CSV", async () => {
    expect(await generateCsv(LABEL_CLASSES_DATA)).toMatchSnapshot();
  });

  it("Triggers download of the CSV file", async () => {
    window.URL.createObjectURL = jest.fn();
    const createElementOriginal = document.createElement.bind(document);
    const onDownload = jest.fn();
    jest
      .spyOn(document, "createElement")
      .mockImplementation((name, options) => {
        const element = createElementOriginal(name, options);
        if (name === "a") {
          element.click = onDownload;
        }
        return element;
      });
    await exportDatasetClasses(DATASET_SLUG, LABEL_CLASSES_DATA);
    await waitFor(() => {
      expect(window.URL.createObjectURL).toHaveBeenCalledTimes(1);
      expect(onDownload).toHaveBeenCalledTimes(1);
    });
  });
});
