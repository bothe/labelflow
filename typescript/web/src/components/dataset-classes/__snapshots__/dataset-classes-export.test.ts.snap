// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelClasses CSV export Generates proper CSV 1`] = `
"Test Class With Labels 1
\\"Test Class With Labels, 2\\"
"
`;
